domain/permission/service/service.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+
+	"github.com/juju/juju/core/permission"
+)
+
+// State describes retrieval and persistence methods for user permission on
+// various targets.
+type State interface {
+	// CreatePermission gives the user access per the provided spec.
+	CreatePermission(ctx context.Context, spec UserAccessSpec) (permission.UserAccess, error)
+
+	// DeletePermission removes the given user's access to the given target.
+	DeletePermission(ctx context.Context, subject names.UserTag, target names.Tag) error
+
+	// UpdatePermission updates the user's access to the given target to the
+	// given access.
+	UpdatePermission(ctx context.Context, subject names.UserTag, target names.Tag, access permission.Access) error
+
+	// ReadUserAccessForTarget returns the user access for the given user on
+	// the given target.
+	ReadUserAccessForTarget(ctx context.Context, subject names.UserTag, target names.Tag) (permission.UserAccess, error)
+
+	// ReadUserAccessLevelForTarget returns the user access level for the
+	// given user on the given target.
+	ReadUserAccessLevelForTarget(ctx context.Context, subject names.UserTag, target names.Tag) (permission.Access, error)
```

Comment:
I'm wondering if these can be combined in the future as UserAccess contains the Access value. How permissions are held is not cohesive across all targets today.